### DISCUSSION TITLE
Pipeline processing features

### DISCUSSION THREAD

**Discussion by JoshClose** [JoshClose/CsvHelper#1738] [Ideas]
@StefanBertels has mentioned using a pipeline to process records. I came across this when converting the docs site over to [Statiq](https://statiq.dev/framework/pipelines/) as it's heavily used there. It makes modifying the behavior of the system really simple. `virtual` methods/members with some `protected` isn't cutting it since moving config and state internal to the classes. Making more things `protected` just causes a lot or problems.

A pipeline would allow any part of the system to easily be changed, added, or removed. It would also give a lot more visibility into how the system behaves.

What are some features that people would like to see in the system? I'd like to make sure what is built can handle them.

I'll start a list and add things as they come in.

1. Side effect free. This will allow for parallelism in processing fields, or possibly even rows.
1. Pipeline steps are just `Func`s or `delegate`s to allow simple lambda expressions as a step. This should allow a user to create their own re-usable library to plug in.

**Comment by TimSirmovics** [comment]
These are some of the features I have written around the current system:

1. On any sort of error with a field, record all information and continue parsing. All errors are available at the end.
2. New attributes and extensions to current attributes for more parsing control.
3. Mapping a multi record file from an object model that contains a top level object, and property references or lists for relations.

Not sure how these would all translate to using a new Pipeline model, but happy to provide code samples if required.
If you start development of Pipelines on another branch I can try and port my code over and provide more specific feedback and perhaps some PR's.

**Reply by JoshClose** [reply]
If you have any thoughts on the current mapping system, let me know. It was designed a long time ago to work similar to FluentNhibernate. There is probably a better way. I've been looking at the same thing for so long that it's hard to see past it.

**Reply by TimSirmovics** [reply]
My goal was to have all specifications of the file defined on the class, so that I can create my object model, annotate it, and then pass that to the library wrapper around CsvHelper. This way we don't have to define the object model and configuration separately. This was done using Attributes, conventions based on the types of properties, and for multi record files relationship properties between the classes.
This way suited us as we often have to implement new file formats so it was the least amount of work to define a new file type with all of its associated configuration.
We only used the fluent interface inside the wrapper library to apply configuration, and there was no real issues with it there.
